server/message_dispatch.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) 2015 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"strings"
+	"sync"
+)
+
+type MessageDispatch struct {
+	input       chan string
+	output      map[string][]chan string
```

Comment:
Hi @koziol,

> As for using Conds, I tend to prefer high-level abstractions (like chan) over low-level primitives (like Cond+Mutex+buffer).

What makes you think, Cond is low level abstraction, while chan is high-level? chan is builtin, while Cond is standard library, which would rather point to the opposite conclusion. In any case, we're talking about:

- Mutex
- map of lists for grouping chans
- push list in wait method
- for loop in broadcast method

vs.

- Mutex
- map of Conds

If you have a look at https://golang.org/src/sync/cond.go?s=1407:1428#L43 , there is actually a similar semantics to what you're proposing, except it's already abstracted.

> As far as I understand, what you're suggesting is actually encapsulating a Cond (plus helper variables) into a chan.

I'm not sure, I understand this sentence. Could you clarify? We're interested in waking up groups of goroutines. What we're doing now, is waiting on a channel created specifically for receiving single notification signal and not caring about it's content. If anything, I think it's exactly the reverse case, where we're trying to force channels to behave like Cond.

> What's more, when designing a prototype with manual synchronization we've came up with a code that was hard to understand and analyze. The problem were sequences of interleaved lock-unlock calls on different objects.

I think I still don't see a problem. I've done small demo, which shows how straight-forward it is to use https://play.golang.org/p/I5LhD7oyjy . Could you check this out and explain what's the problem with this approach?